src/commands/azureServiceOperators/ui/azureservicebrowser.ts

```diff
@@ -16,21 +21,34 @@ export async function AzureServiceBrowser(explorer: k8s.ClusterExplorerV1): Prom
 }
 
 async function allServiceKinds(): Promise<AzureServiceKind[] | undefined> {
+    const apiResult = await getAPIResourceCommandResult();
+
+    if (apiResult.succeeded) {
+        return apiResult.result;
+    } else if (apiResult.error !== ASOInstallation.ASONotInstalled) {
+        vscode.window.showWarningMessage(apiResult.error);
+    }
+    return undefined;
+}
+
+async function getAPIResourceCommandResult(): Promise<Errorable<AzureServiceKind[]>> {
     const kubectl = await k8s.extension.kubectl.v1;
 
     if (!kubectl.available) {
-        vscode.window.showWarningMessage(`Kubectl is unavailable.`);
-        return undefined;
+        return { succeeded: false, error: `Kubectl is unavailable.` };
     }
     const asoAPIResourceCommandResult = await kubectl.api.invokeCommand("api-resources --api-group azure.microsoft.com --no-headers");
-    if (!asoAPIResourceCommandResult || (asoAPIResourceCommandResult.code !== 0 && !asoAPIResourceCommandResult.stdout)) {
-        vscode.window.showWarningMessage(`Azure Service Operator api-resources command failed with following error: ${asoAPIResourceCommandResult?.stderr}.`);
-        return undefined;
-    }
 
-    const treeResourceItems = asoAPIResourceCommandResult.stdout.split("\n").map((line: string) => line.trim()).filter((line: string | any[]) => line.length > 0);
-
-    return treeResourceItems.map((item: string) => parseServiceResource(item)!);
+    if (!asoAPIResourceCommandResult) { // Fail to invoke command.
+        return { succeeded: false, error: `Azure Service Operator api-resources failed to invoke command.` };
+    } else if (asoAPIResourceCommandResult.code !== 0 && !asoAPIResourceCommandResult.stdout) { // Error result and Faild execution.
+        return { succeeded: false, error: `Azure Service Operator api-resources command failed with following error: ${asoAPIResourceCommandResult?.stderr}.` };
+    } else if (asoAPIResourceCommandResult.code === 0 && !asoAPIResourceCommandResult.stdout) { // No ASO installed.
+        return { succeeded: false, error: ASOInstallation.ASONotInstalled };
+    } else { // success sceanrio.
+        const treeResourceItems = asoAPIResourceCommandResult.stdout.split("\n").map((line: string) => line.trim()).filter((line: string | any[]) => line.length > 0);
```

Comment:
Don't get how `line` would be anything other than a `string` in the `filter` clause; why do you need `| any[]` here?